src/data/nat/count.lean

```diff
@@ -0,0 +1,542 @@
+/-
+Copyright (c) 2021 Vladimir Goryachev. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Vladimir Goryachev, Kyle Miller, Scott Morrison, Eric Rodriguez
+-/
+import data.list.basic
+import data.nat.prime
+import order.order_iso_nat
+import set_theory.fincard
+
+/-!
+# Counting on ℕ
+
+This files defines ways to get basic properties of a predicate on ℕ, such as "how many numbers
+under `k` satisfy the predicate" and "what is the `n`th number that satisifies this predicate".
+We define these as two functions, `count` and `nth`, that answer these questions, and prove
+the expected theorems about them.
+
+## Main definitions
+
+* `count p n`: The number of naturals `k < n` such that `p k`.
+* `nth p n`: The `n`-th natural `k` (zero-indexed) such that `p k`. If there is no
+  such natural (that is, `p` is true for at most `n` naturals), then `nth p n = 0`.
+
+## Main results
+
+* `set.infinite.order_iso_nat`: An infinite set of natural numbers is order-isomorphic to the
```

Comment:
Documentation error?  I could not find a result in the file of this name.

Reply:
The original reason was essentially just that we found it during writing this file :/ I assume this also got eschewed because of the order iso existing already

Reply:
First, even if the finite case is in some sense uninteresting, it is stiil covered by the definition of `nth`, but not by `nat.subtype.order_iso_of_nat`. As some of the theorems are specifically about the finite case, and some are valid in both the infinite and finite case, it would not be possible to formulate them in the same generality (or at all) using the `nat.subtype.order_iso_of_nat`.

Second, the definiton of `nat.subtype.order_iso_of_nat`, is kind of awkward. Rather saying "the least number satisfying p, which is greater than the previous one", it says, after you expand it, "the previous number plus one plus (the least number such that if we add to it the previous number plus one, we get a number satisfying p)". In the end, this is equivalent to asking the result to be greater than the previous one, as we are working with the natural numbers, but the definition of `nth` is still cleaner.

Third, the `nth` name is more intuitive, and closer to how we think about the function, not to the underlying abstraction.

Considering the three points, it probably makes sense to redefine `nat.subtype.order_iso_of_nat` in terms of `nth`, rather than vice versa, and this suggestion was on the table during the zulip discussions. However, I am not sure it makes sense to do it in this PR, as the guideline is to keep the PRs rather small, and we already have a bunch of proofs here. Also, this `nat.subtype.order_iso_of_nat` seems closer to foundational aspects of Lean, and personally I do not understand those areas well enough to be confident about changing them, and would welcome advice on whether it is reasonable.

Reply:
I don't agree here.  I think the first point you make is reasonable.  But regarding the second point, it doesn't matter how things are defined; your second point is just an argument in favour of providing the current _definition_ of `nth` as a _lemma_ about `order_iso_of_nat`.  And your third point is just an argument in favour of (possibly) renaming `order_iso_of_nat` to `nth`, although we should proceed carefully here because it doesn't really follow mathlib naming conventions (but might be a nice enough name to justify breaking them).

Having two definitions of (nearly) the same thing can lead to huge code duplication, and it really seems unnecessary here.

Reply:
It might be nice to be able to deal with finite sets or sets not known to be infinite, so having `nth` makes sense. `order_iso_of_nat` has the additional isomorphism structure, so that makes sense as a definition too. As I see it, no one definition will be able to support both finite sets and the order isomorphism. 

I have left a note in the module docstring referencing this discussion, as well as notes in the docstrings of both `subtype.order_iso_of_nat` and `nth` referencing each other. Hopefully this will help future contributors decide which location is best for their lemmas.